redis-cli: Accept commands in subscribed mode

zuiderkwast opened the pull request:
The message "Reading messages... (press Ctrl-C to quit)" is replaced by "Reading messages... (press Ctrl-C to quit or any key to type command)".

This allows users to subscribe to more channels, to try out UNSUBSCRIBE and to combine pubsub with other features such as push messages from client tracking.

The "Reading messages" info message is displayed in the bottom of the output in a distinct style and moves downward as more messages appear. When any key is pressed, the info message is replaced by the prompt with for entering commands. After entering a command and the reply is displayed, the "Reading messages" info messages appears again. This is added to the repl loop in redis-cli and in the corresponding place for non-interactive mode.

An indication "(subscribed mode)" is included in the prompt when entering commands in subscribed mode.

Also:

* Fixes a problem that UNSUBSCRIBE hanged when used with RESP3 and push callback, without first entering subscribe mode. It hanged because UNSUBSCRIBE gets one or more push replies but no in-band reply.

* Exit subscribed mode after RESET.

bjosv commented:
Very nice!
As an early tester (and probably lack of insights) I've seen that every second PING response after a subscribe is not printed, like below. Not sure if this is a legacy problem.
```
./src/redis-cli -3 -p 7000
127.0.0.1:7000> subscribe mychan
Reading messages... (press Ctrl-C to quit or enter to type command)
1) "subscribe"
2) "mychan"
3) (integer) 1
PING                                                                  <--- No PONG printed after this
127.0.0.1:7000(subscribed mode)> PING
PONG
Reading messages... (press Ctrl-C to quit or enter to type command)
PING                                                                  <--- No PONG printed after this
127.0.0.1:7000(subscribed mode)> 
```

zuiderkwast commented:
@bjosv I changed "press enter" to "press any key" to avoid the confusion you experienced. I hope it's more intuitive now.

Added tests for TTY and non-TTY.

zuiderkwast commented:
Some screenshots:
![redis-cli-subscribed-mode-1](https://user-images.githubusercontent.com/273886/223200561-c3c1d661-5e65-41b3-a3d8-697a49f55ec2.png)
![redis-cli-subscribed-mode-2](https://user-images.githubusercontent.com/273886/223200977-e9097944-5b2d-4545-b000-73e37975936d.png)
![redis-cli-subscribed-mode-3](https://user-images.githubusercontent.com/273886/223200995-8374b7ed-8154-44f4-8ed3-a6064508f11b.png)
![redis-cli-subscribed-mode-4](https://user-images.githubusercontent.com/273886/223201004-2f9d21c8-9011-46bf-b867-7fb741d21f3a.png)

bjosv reviewed (commented):
I seems that my terminal is not showing the "Reading messages.. .." anymore.
![image](https://user-images.githubusercontent.com/60651423/223664779-69d87d55-5c65-4c0e-b3c4-f4bf00ed7590.png)

After some investigations my `TERM=xterm-256color` using `solarized dark` don't like
the color `\033[1;90m` ("Bold, bright color.").

Maybe the black color wont work for other dark mode users either.

zuiderkwast commented:
@bjosv The color 90 "bright black (gray)" is the same color as already used for command hints. I guess you don't see hints either then?

bjosv commented:
Nope, seems that I'm missing out of some good stuff :flushed:
Well, you last change works fine for me even not seeing the text, it's intuitive.

oranagra commented:
wow.. that's a lot of CLI code (hope i'll find time to review it).
btw, maybe one of you is brave enough to review the redis-cli portion of https://github.com/redis/redis/pull/10515? and then i can merge it.

oranagra commented on src/redis-cli.c:
```diff
@@ -1892,8 +2027,23 @@ static int cliSendCommand(int argc, char **argv, long repeat) {
                 cliRefreshPrompt();
+            } else if (!strcasecmp(command,"hello")) {
+                if (config.last_cmd_type == REDIS_REPLY_MAP)
+                    config.resp3 = 2;
```
i may be out of context, but isn't config.resp3 suppose to already be set to 1 or 2?
in which case overriding it with 2 may be wrong?
maybe set it to 2 (or actually 1?) only if it was 0 before?

zuiderkwast replied:
Good points. I assumed users only type HELLO if they want to switch to RESP3, but you're right it can be called even if RESP3 is already active.

The values 1 (explicit) and 2 (implicit) seemed vague to me. Looking at how the values are used, I realize 2 means it can be overridden by -2 (for example `--json -2`  means use RESP2) and also it is not an error if the initial `HELLO 3` fails if resp3 == 2, which I suppose it to be able support older Redis versions where RESP3 was not implemented (and HELLO is an unknown command).

Here after HELLO 3 has returned a map, we know that we have RESP3 so setting it to 1 seems right. I think we can do it regardless of the previous value (0, 1 or 2) but I can do it only `if (config.resp3 == 0)` if you think it's better.

oranagra replied:
that code also run when we get the response of an implicit HELLO 3 we sent (i.e. when user just passed `--json`), right?

looking closer, i see this variable is actually tied to command line arguments, and controls which HELLO we'll send when connecting / reconnecting.
so i wonder if at all we should modify it post startup (also for RESET, which already modifies it).
maybe instead we should have another variable that's indicating the current connection's RESP mode (separate from the user's intention when starting redis-cli, which controls the (re)connection?

zuiderkwast replied:
> that code also run when we get the response of an implicit HELLO 3 we sent (i.e. when user just passed `--json`), right?

No. This code only runs on user input HELLO.

> looking closer, i see this variable is actually tied to command line arguments, and controls which HELLO we'll send when connecting / reconnecting.
> so i wonder if at all we should modify it post startup (also for RESET, which already modifies it).
> maybe instead we should have another variable that's indicating the current connection's RESP mode (separate from the user's intention when starting redis-cli, which controls the (re)connection?

Is it really that important? The only difference is that you get RESP3 again after reconnect, rather than being reset to what you had when connecting, which is probably what the user wants anyway if they've explicitly issued a HELLO command.

oranagra replied:
well, ok, i suppose this is the path of least surprise.

zuiderkwast replied:
One scenario I can think of: A new redis-cli connects to an old redis-server without RESP3 support. The user specifies e.g. `--json` so config.resp3 == 2. This doesn't cause cliSwitchProto to fail, so the user ends up connected with RESP2. The problem then is the check for pubsub messages.

```C
int isPubsubPush(redisReply *r) {
    if (r == NULL ||
        r->type != (config.resp3 ? REDIS_REPLY_PUSH : REDIS_REPLY_ARRAY) ||
...
```

To support this scenario, I guess we need to know if we actually have RESP3 right now or not.

zuiderkwast replied:
Now I added a `config.current_resp3` that keeps track of the current state, as you suggested some hours ago.

oranagra reviewed (approved):
LGTM.
hope i'm not missing any corner cases.
please run some use cases (interactive, non-interactive, piped from stdin) via valgrind.
@yossigo please take a quick look and state if you see any problem with it.

yossigo commented:
This looks really nice! @zuiderkwast Did you consider actually using the first keypress that breaks out of the receive loop as input? Intuitively I just started typing but then noticed the first character is "lost".

zuiderkwast commented:
@yossigo The "any key" can be used as input if we change the following in linenoise. I don't know if it has any drawbacks. Probably it will just make input available that otherwise would be lost if the user is typing too fast. Is it worth it? WDYT?

```diff
diff --git a/deps/linenoise/linenoise.c b/deps/linenoise/linenoise.c
index 1b01622c5..c6c496e34 100644
--- a/deps/linenoise/linenoise.c
+++ b/deps/linenoise/linenoise.c
@@ -257,7 +257,7 @@ static int enableRawMode(int fd) {
     raw.c_cc[VMIN] = 1; raw.c_cc[VTIME] = 0; /* 1 byte, no timer */
 
     /* put terminal in raw mode after flushing */
-    if (tcsetattr(fd,TCSAFLUSH,&raw) < 0) goto fatal;
+    if (tcsetattr(fd,TCSANOW,&raw) < 0) goto fatal;
     rawmode = 1;
     return 0;
```

oranagra commented:
i've tested it and i see two things that could maybe improved:
1. after hitting a key and getting a prompt, if i type enter i get a new line without a prompt, another enter gives me a prompt again, and a 3rd one give me the "reading messages" line again.
2. if i start typing and don't hit enter, i won't get any push message until i hit enter (even if i delete what i started typing, i won't get the push until i post a command or hit enter).
3. if some push was received while i type, when i do hit enter, it's hard to distinguish between the post message and the command response (they look similar)

zuiderkwast commented:
> i've tested it and i see few things that could maybe improved:
> 
> 1. after hitting a key and getting a prompt, if i type enter i get a new line without a prompt, another enter gives me a prompt again, and a 3rd one give me the "reading messages" line again.

This must have been after the last commit (Use the "any key" as input). The enter goes to the prompt, which goes back to "reading messages..." I don't really know what to do in this case, without breaking the abstraction of linenoise. Any ideas?

> 2. if i start typing and don't hit enter, i won't get any push message until i hit enter (even if i delete what i started typing, i won't get the push until i post a command or hit enter).

That's expected. When the prompt is visible, linenoise is doing a blocking read() on stdin. It doesn't use select() to detect input on the socket here, since this is a separate library which is only concerned about line editing. Do you think I should break the abstraction of this lib and read push messages during line editing, or inject some way to do it, using ifdefs or something like that?

> 3. if some push was received while i type, when i do hit enter, it's hard to distinguish between the post message and the command response (they look similar)

Right. This is easier to solve. How about using a different color for push messages? Cyan, magenta...

oranagra commented:
1. by "solved" you mean a future commit? i do think i tested the latest.
2. maybe we can leave that aside, but we do need it to be clear to the user that he's currently not watching for pushes (e.g. if he deleted what he started to type), and what he should do in order to listen again.
3. we don't use colors for other server outputs. maybe add some indentation or some `>` prefix? (we better do that only in interactive mode)

zuiderkwast commented:
> * by "solved" you mean a future commit? i do think i tested the latest.

Exactly :-) Sorry for the confusion.

> * maybe we can leave that aside, but we do need it to be clear to the user that he's currently not watching for pushes (e.g. if he deleted what he started to type), and what he should do in order to listen again.

Isn't it clear enough that when the "Reading messages..." is displayed, push messages are displayed continuously, but when the prompt is displayed, it can't be assumed that they are?

> * we don't use colors for other server outputs. maybe add some indentation or some `>` prefix? (we better do that only in interactive mode)

OK, for RESP3 we can use a prefix for all push messages. I think that's a good idea.

For RESP2 though, the user can't really tell what's a "push" and what's a reply. There is no difference. That's why only a few commands are allowed in RESP2 pubsub mode. The user can disambiguate by looking at the replies.

oranagra commented on src/redis-cli.c:
```diff
@@ -1345,6 +1345,7 @@ static sds cliFormatReplyTTY(redisReply *r, char *prefix) {
             char numsep;
             if (r->type == REDIS_REPLY_SET) numsep = '~';
             else if (r->type == REDIS_REPLY_MAP) numsep = '#';
+            else if (r->type == REDIS_REPLY_PUSH) numsep = '>';
```
aren't we introducing a breaking change (for scripts that work with redis-cli)?
i thought to do that only in interactive mode, but actually that would be confusing (someone experimenting with interactive in order to know how to write his script).
@yossigo please advise.

zuiderkwast replied:
This output is only for TTYs, so script shouldn't be affected unless they're trying to simulate a TTY (like we do in the TTY tests).

zuiderkwast replied:
It seems `redis-cli --no-raw` in scripts is affected, so it's breaking change for this use case.

zuiderkwast replied:
All other types have their own prefix, so it's odd that push doesn't. Push shares the prefix with array (`1)`, `2)` etc.).

OTOH, this is not coupled with this PR so I can move it to a separate PR.

oranagra replied:
I spoke to Yossi, let's drop this `>` change, and add it in 8.0

oranagra commented:
> Isn't it clear enough that when the "Reading messages..." is displayed, push messages are displayed continuously, but when the prompt is displayed, it can't be assumed that they are?

i suppose you're right. testing the new last version and knowing that seems that it works right.

looking for the `>` i now realize that i tested pubsub and not push, so the `>` didn't help me, and the responses are still mixed.
```
127.0.0.1:6379(subscribed mode)> ping
1) "message"
2) "asdf"
3) "asdfasdf111asdf"
1) "pong"
2) ""
```

another interesting thing i just noticed, that while in that subscribed mode, i'm getting another line (`2) ""`), can you explain it?

zuiderkwast commented:
> looking for the `>` i now realize that i tested pubsub and not push, so the `>` didn't help me, and the responses are still mixed.

Well, try HELLO 3 and then SUBSCRIBE x y z, then you'll see the `>`.

> another interesting thing i just noticed, that while in that subscribed mode, i'm getting another line (`2) ""`), can you explain it?

It is explained on the page https://redis.io/commands/ping/: "If the client is subscribed to a channel or a pattern, it will instead return a multi-bulk with a "pong" in the first position and an empty bulk in the second position, unless an argument is provided in which case it returns a copy of the argument."

oranagra commented:
@zuiderkwast maybe you can have a look at a sporadic error from the new test of this PR:
https://github.com/redis/redis/actions/runs/4585861784/jobs/8098245787#step:4:8565
```
  *** [err]: Interactive CLI: Subscribed mode in tests/integration/redis-cli.tcl
  Expected '1) "unsubscribe"
  2) "ch1"
  3) (integer) 2
  1) "unsubscribe"
  2) "ch2"
  3) (integer) 1
  Reading messages... (press Ctrl-C to quit or any key to type command)
  ' to be equal to '' (context: type eval line 24 cmd {assert_equal $erase$unsub1$unsub2$reading  [run_command $fd "unsubscribe ch1 ch2"]} proc ::test)
```
maybe a timing issue (freebsd CI is notorious for being slow)

zuiderkwast commented:
I can't tell why this failed. This `run_command` is like every other command in this test suite. We could try to increase the hard-coded numbers in `read_cli`, but that would make all checks slower in this test suite.

```tcl
    proc read_cli {fd} {
        set ret [read $fd]
        while {[string length $ret] == 0} {
            after 10
            set ret [read $fd]
        }

        # We may have a short read, try to read some more.
        set empty_reads 0
        while {$empty_reads < 5} {
            set buf [read $fd]
            if {[string length $buf] == 0} {
                after 10
                incr empty_reads
            } else {
                append ret $buf
                set empty_reads 0
            }
        }
        return $ret
    }
```
